File: ios/RecipeClipper/Data/Model/GroceryDecisions.swift
import Foundation

/// The on-device model's help with the grocery list (#99, over #104's `DecisionRule`), ported
/// from the Kotlin `GroceryDecisions` (see there). The model never writes a number: an answer
/// only regroups lines or drops trailing text; `GroceryCombiner`'s exact rules decide totals.
enum GroceryDecisions {

    /// A line cut into its `core` ("2 eggs") and the `trailing` text after the name (", beaten").
    struct Split: Equatable {
        let core: String
        let trailing: String
    }

    private static let separator = JRegex(#"[,;(]|\s[-–—]+(?=\s|$)"#)

    /// The line cut at the first comma, semicolon, bracket or dash whose left side has an
    /// `IngredientName`, or nil (no such cut, a language without spaces or name-first, or
    /// trailing text holding a digit: a figure is never ignored).
    static func split(_ line: String, words: LanguageWords) -> Split? {
        if !words.spaced || IngredientScaler.patterns(words).amountAfterName { return nil }
        for m in separator.findAll(line) {
            let core = trimEnd(line.u16Substring(0, m.start))
            if IngredientName.of(core, words: words) == nil { continue }
            let trailing = line.u16Substring(from: m.start).kTrimmed
            if trailing.utf16.count < 2 || trailing.unicodeScalars.contains(where: { CharacterSet.decimalDigits.contains($0) }) {
                return nil
            }
            return Split(core: core, trailing: trailing)
        }
        return nil
    }

    private static func trimEnd(_ s: String) -> String {
        var s = s
        while let last = s.last, last.isWhitespace { s.removeLast() }
        return s
    }

    /// The text the combiner reads `item` as: its core once its trailing text is a note or junk.
    static func effectiveText(_ item: GroceryItem, decisions: Decisions) -> String {
        guard let words = LanguageWords.forTag(item.language), let split = split(item.text, words: words) else { return item.text }
        return decisions.ignorableTrailing(split.trailing, language: words.language) ? split.core : item.text
    }

    /// `item`'s name as the combiner groups it, or nil.
    static func name(_ item: GroceryItem, decisions: Decisions) -> String? {
        LanguageWords.forTag(item.language).flatMap { IngredientName.of(effectiveText(item, decisions: decisions), words: $0) }
    }

    private static func aislesMeet(_ a: GroceryItem, _ b: GroceryItem) -> Bool {
        a.aisle == b.aisle || a.aisle == .other || b.aisle == .other
    }

    /// Close names on two lines in one language whose aisles could meet.
    static func samePairs(_ items: [GroceryItem], decisions: Decisions) -> [DecisionQuestion] {
        let named = items.compactMap { item in name(item, decisions: decisions).map { (item, $0) } }
        var out: [DecisionQuestion] = []
        for i in named.indices {
            for j in named.indices where j > i {
                let (a, na) = named[i], (b, nb) = named[j]
                guard a.language == b.language, na != nb, aislesMeet(a, b), let words = LanguageWords.forTag(a.language) else { continue }
                let q = DecisionQuestion.sameGrocery(na, nb, language: words.language)
                if DecisionCandidates.close(na, nb, words: words) && !out.contains(q) { out.append(q) }
            }
        }
        return out
    }

    /// Trailing text on a line whose core names what another line names, where they don't add up as written.
    static func trailingTexts(_ items: [GroceryItem], decisions: Decisions) -> [DecisionQuestion] {
        var out: [DecisionQuestion] = []
        for item in items {
            guard let words = LanguageWords.forTag(item.language), let split = split(item.text, words: words),
                  let core = IngredientName.of(split.core, words: words) else { continue }
            let partner = items.contains { other in
                other.id != item.id && other.language == item.language && aislesMeet(item, other)
                    && (IngredientName.of(other.text, words: words) == core || name(other, decisions: decisions) == core)
                    && GroceryCombiner.combine([item.text, other.text], words: words) == nil
            }
            let q = DecisionQuestion.trailingText(split.trailing, language: words.language)
            if partner && !out.contains(q) { out.append(q) }
        }
        return out
    }

    /// Where answers that just landed (`fresh`) file lines from Other (see the Kotlin).
    static func filing(_ items: [GroceryItem], fresh: Set<DecisionQuestion>, decisions: Decisions) -> [Aisle: [Int64]] {
        var moves: [(Int64, Aisle)] = []
        for item in items where item.aisle == .other {
            guard let words = LanguageWords.forTag(item.language) else { continue }
            if let split = split(item.text, words: words),
               fresh.contains(.trailingText(split.trailing, language: words.language)),
               decisions.ignorableTrailing(split.trailing, language: words.language) {
                let aisle = Aisles.of(split.core, words: words)
                if aisle != .other { moves.append((item.id, aisle)); continue }
            }
            guard let name = name(item, decisions: decisions) else { continue }
            let partner = items.first { other in
                guard other.aisle != .other, other.language == item.language,
                      let n = GroceryDecisions.name(other, decisions: decisions) else { return false }
                return fresh.contains(.sameGrocery(name, n, language: words.language))
                    && decisions.sameGrocery(name, n, language: words.language)
            }
            if let partner { moves.append((item.id, partner.aisle)) }
        }
        return Dictionary(grouping: moves, by: { $0.1 }).mapValues { $0.map(\.0) }
    }
}

File: app/src/main/java/com/example/recipeclipper/data/model/GroceryDecisions.kt
package com.example.recipeclipper.data.model

/**
 * The on-device model's help with the grocery list (#99, over #104's [DecisionRule]): which
 * lines' close names to ask about, which text after an ingredient to ask about, and how a
 * line reads once answered. Pure, both platforms. The model never writes a number: a definite
 * answer only lets [GroceryCombiner] group lines, or read a line without its trailing text,
 * and the combiner's own exact rules still decide every total.
 */
object GroceryDecisions {

    /** A line cut into its [core] ("2 eggs") and the [trailing] text after the name (", beaten"). */
    data class Split(val core: String, val trailing: String)

    private val SEPARATOR = Regex("""[,;(]|\s[-–—]+(?=\s|$)""")

    /**
     * The line cut at the first comma, semicolon, bracket or dash whose left side has an
     * [IngredientName], or null: no such cut, a language written without spaces or name-first,
     * or trailing text holding a digit (a figure is never ignored, whatever the model says).
     */
    fun split(line: String, words: LanguageWords): Split? {
        if (!words.spaced || IngredientScaler.patterns(words).amountAfterName) return null
        for (m in SEPARATOR.findAll(line)) {
            val core = line.substring(0, m.range.first).trimEnd()
            if (IngredientName.of(core, words) == null) continue
            val trailing = line.substring(m.range.first).trim()
            if (trailing.length < 2 || trailing.any { it.isDigit() }) return null
            return Split(core, trailing)
        }
        return null
    }

    /** The text [GroceryCombiner] reads [item] as: its core once its trailing text is definitely a note or junk. */
    fun effectiveText(item: GroceryItem, decisions: Decisions): String {
        val words = LanguageWords.forTag(item.language) ?: return item.text
        val split = split(item.text, words) ?: return item.text
        return if (decisions.ignorableTrailing(split.trailing, words.language)) split.core else item.text
    }

    /** [item]'s name as the combiner groups it, or null. */
    fun name(item: GroceryItem, decisions: Decisions): String? =
        LanguageWords.forTag(item.language)?.let { IngredientName.of(effectiveText(item, decisions), it) }

    // Two lines can share a row only in one aisle; one in Other may be filed beside the other.
    private fun aislesMeet(a: GroceryItem, b: GroceryItem) =
        a.aisle == b.aisle || a.aisle == Aisle.OTHER || b.aisle == Aisle.OTHER

    /**
     * The "same thing?" questions worth asking: two lines in one language whose names differ,
     * are [DecisionCandidates.close], and whose aisles could meet.
     */
    fun samePairs(items: List<GroceryItem>, decisions: Decisions): List<DecisionQuestion> {
        val named = items.mapNotNull { item -> name(item, decisions)?.let { item to it } }
        val out = LinkedHashSet<DecisionQuestion>()
        for (i in named.indices) for (j in i + 1 until named.size) {
            val (a, na) = named[i]
            val (b, nb) = named[j]
            if (a.language != b.language || na == nb || !aislesMeet(a, b)) continue
            val words = LanguageWords.forTag(a.language) ?: continue
            if (DecisionCandidates.close(na, nb, words)) out += DecisionQuestion.sameGrocery(na, nb, words.language)
        }
        return out.toList()
    }

    /**
     * The trailing-text questions worth asking: a line whose core names what another line in
     * its language names, where the two don't already add up as written.
     */
    fun trailingTexts(items: List<GroceryItem>, decisions: Decisions): List<DecisionQuestion> {
        val out = LinkedHashSet<DecisionQuestion>()
        for (item in items) {
            val words = LanguageWords.forTag(item.language) ?: continue
            val split = split(item.text, words) ?: continue
            val core = IngredientName.of(split.core, words) ?: continue
            val partner = items.any { other ->
                other.id != item.id && other.language == item.language && aislesMeet(item, other) &&
                    (IngredientName.of(other.text, words) == core || name(other, decisions) == core) &&
                    GroceryCombiner.combine(listOf(item.text, other.text), words) == null
            }
            if (partner) out += DecisionQuestion.trailingText(split.trailing, words.language)
        }
        return out.toList()
    }

    /**
     * Where answers that just landed ([fresh]) file lines from Other: beside a line in another
     * aisle now decided the same, or where the table puts a line's core once its trailing text
     * is a note or junk. Only lines in Other move, and only on a fresh answer, so an aisle the
     * user chose stands.
     */
    fun filing(items: List<GroceryItem>, fresh: Set<DecisionQuestion>, decisions: Decisions): Map<Aisle, List<Long>> {
        val moves = LinkedHashMap<Long, Aisle>()
        for (item in items.filter { it.aisle == Aisle.OTHER }) {
            val words = LanguageWords.forTag(item.language) ?: continue
            val split = split(item.text, words)
            if (split != null && DecisionQuestion.trailingText(split.trailing, words.language) in fresh &&
                decisions.ignorableTrailing(split.trailing, words.language)
            ) {
                Aisles.of(split.core, words).takeIf { it != Aisle.OTHER }?.let { moves[item.id] = it }
            }
            val name = name(item, decisions) ?: continue
            if (item.id in moves) continue
            items.firstOrNull { other ->
                other.aisle != Aisle.OTHER && other.language == item.language &&
                    name(other, decisions)?.let { n ->
                        DecisionQuestion.sameGrocery(name, n, words.language) in fresh &&
                            decisions.sameGrocery(name, n, words.language)
                    } == true
            }?.let { moves[item.id] = it.aisle }
        }
        return moves.entries.groupBy({ it.value }, { it.key })
    }
}
